File: player.rb
class Player
	attr_reader :username, :socket, :level, :max_health, :health, :attack, :defense, :total_experience, :remaining_experience_required_levelup, :total_experience_required_levelup
	attr_accessor :location, :defending, :in_fight, :player_turn, :target_player
	EXP_GROWTH_RATE = 1.1

	def initialize(username, socket, location)
		@username = username
		@socket = socket
		@location = location
		@level = 1
		@max_health = 30
		@health = 30
		@attack = 15
		@defense = 5
		@defending = false
		@in_fight = false
		@player_turn = true
		@total_experience = 0.0
		@remaining_experience_required_levelup = 100.0
		@total_experience_required_levelup = 100.0
		@target_player = nil
	end
	def level_up
		@level += 1
		@max_health += 5
		@health = @max_health
		@attack += 2
		@defense += 1
	end
	def take_damage(damage)
	    damage_taken = [0, damage - @defense].max
	    damage_taken /= 2 if @defending
	    @health -= damage_taken
	    @defending = false
	    return damage_taken 
	end
	def alive?
		@health > 0
	end
	def heal
		heal_amount = (@max_health * 0.20).to_i
  		@health += heal_amount
  		@health = [@health, @max_health].min
  		return heal_amount
	end
	def full_heal
		@health = @max_health
	end
	def add_experience(level)
		experience_to_add = (level.to_f / @level) * 10 * (1 + (level * 0.1))
		@total_experience += experience_to_add
		@remaining_experience_required_levelup -= experience_to_add
		if @remaining_experience_required_levelup <= 0
			excess_experience = @remaining_experience_required_levelup.abs
			level_up()
			@total_experience_required_levelup = (@total_experience_required_levelup * EXP_GROWTH_RATE)
			@remaining_experience_required_levelup = @total_experience_required_levelup - excess_experience
		end
		return experience_to_add
	end
end

File: player.js
class Player {
	#username;
	#socket;
	#level;
	#maxHealth;
	#health;
	#attack;
	#defense;
	#totalExperience;
	#remainingExperienceRequiredLevelup;
	#totalExperienceRequiredLevelup;
	location;
	defending;
	inFight;
	playerTurn;
	targetPlayer;
	static EXP_GROWTH_RATE = 1.1;

	constructor(username, socket, location){
		this.#username = username;
		this.#socket = socket;
		this.location = location;
		this.#level = 1;
		this.#maxHealth = 30;
		this.#health = 30;
		this.#attack = 15;
		this.#defense = 5;
		this.defending = false;
		this.inFight = false;
		this.playerTurn = true;
		this.#totalExperience = 0.0;
		this.#remainingExperienceRequiredLevelup = 100.0;
		this.#totalExperienceRequiredLevelup = 100.0;
		this.targetPlayer = null;
	}

	get username(){ return this.#username; }
	get socket(){ return this.#socket; }
	get level(){ return this.#level; }
	get maxHealth(){ return this.#maxHealth; }
	get health(){ return this.#health; }
	get attack(){ return this.#attack; }
	get defense(){ return this.#defense; }
	get totalExperience(){ return this.#totalExperience }
	get remainingExperienceRequiredLevelup(){ return this.#remainingExperienceRequiredLevelup }
	get totalExperienceRequiredLevelup(){ return this.#totalExperienceRequiredLevelup }

	levelUp(){
		this.#level += 1;
		this.#maxHealth += 5;
		this.#health = this.#maxHealth;
		this.#attack += 2;
		this.#defense += 1;
	}

	takeDamage(damage){
		let damageTaken = Math.max(0, damage - this.#defense);
		if (this.defending){
			damageTaken /= 2;
		}
		this.#health -= damageTaken;
		this.defending = false;
		return damageTaken
	}

	isAlive(){
		return this.#health > 0;
	}

	heal(){
		let healAmount = parseInt(this.maxHealth * 0.20);
		this.#health += healAmount;
		this.#health = Math.min(this.#health, this.#maxHealth);
		return healAmount;
	}

	fullHeal(){
		this.#health = this.#maxHealth;
	}

	addExperience(level){
		let experienceToAdd = (parseFloat(level) / this.#level) * 10 * (1 + (level * 0.1));
		this.#totalExperience += experienceToAdd;
		this.#remainingExperienceRequiredLevelup -= experienceToAdd;
		if(this.#remainingExperienceRequiredLevelup <= 0){
			let excessExperience = Math.abs(this.#remainingExperienceRequiredLevelup);
			this.levelUp();
			this.#totalExperienceRequiredLevelup = this.#totalExperienceRequiredLevelup * Player.EXP_GROWTH_RATE;
			this.#remainingExperienceRequiredLevelup = this.#totalExperienceRequiredLevelup - excessExperience;
		}
		return experienceToAdd;
	}
}
module.exports = Player;